tmt/utils.py

```diff
@@ -1279,6 +1293,42 @@ def copytree(
     return dst
 
 
+def get_full_metadata(fmf_tree_path: str, node_path: str) -> Any:
+    """
+    Get full metadata for a node in any fmf tree
+    Go through fmf tree nodes using given relative node path
+    and return full data as dictionary
+    """
+    return fmf.Tree(fmf_tree_path).find(node_path).data
+
+
+def filter_paths(directory: Path, searching: List[str], files_only: bool = False) -> List[Path]:
+    """
+    Filter files for specific paths we are searching for inside a directory
+    Returns list of matching paths
+    """
+    allfiles = list(directory.rglob('*'))
+    alldirs = [d for d in allfiles if d.is_dir()]
+    allfiles = [f for f in allfiles if not f.is_dir()]
+    allfiles = list(map(str, allfiles))  # type: ignore[arg-type]
+    alldirs = list(map(str, alldirs))  # type: ignore[arg-type]
+    found_paths = []
+
+    for search_string in searching:
+        regex = re.compile(search_string)
+
+        if not files_only:
+            # Search in directories first to reduce amount of copying later
+            matches = list(filter(regex.search, alldirs))  # type: ignore[arg-type]
+            if matches:
+                found_paths += matches
+                continue
+
+        # Search through all files
+        found_paths += list(filter(regex.search, allfiles))  # type: ignore[arg-type]
+    return list(map(Path, found_paths))  # type: ignore[arg-type]
```

Comment:
One quick note: please, add a comment like https://github.com/teemtee/tmt/blob/main/tmt/base.py#L114 explaining why `ignore` is needed, ideally for each of the uses of `type: ignore`. Without such comments, it's hard to review now and revisit in the future should the situation change.

I suppose at least some of them are because of names like `allfiles` referring to different types during their lifetime, but it's hard to guess correctly.

Reply:
you guessed right, will add a comment; I had to add quite a number of type ignores

Reply:
* I'd advise against re-using names for different types, it makes the job harder for mypy. It does not change what mypy thinks about the name, just silences the warning, therefore these waivers tend to multiply. Every time the name is used in the context of new type, mypy would complain. It's generally safer - and more readable, IMHO - to use different names.
* single character names are often frowned upon (I for one don't mind using them, but I recall seeing comments targeting `i` and `s` and others...), therefore I'd recommend avoiding them.
* `iterdir()` gives out the content, no need for listing `*` via `glob()`
* I'd not bother with `Path` => `str` conversion and back

```python
    directory_content = list(directory.iterdir())

    directory_names = [str(path) for directory_content if path.is_dir()]
    file_names = [str(path) for directory_content if not path.is_dir()]

    found_names: List[str] = []
    ...
```

Reply:
iterdir does not list all subdirectories recursively
otherwise, yeah just thought about using different names, will change it so those type ignores are not there, they are not that important and it might even be more readable